### PR TITLE
add docs TypicalLogitsWarper

### LINKED ISSUE #24783: Generate: have an example on each `LogitsProcessor` class docstring
# Context

`.generate()` can be extensively manipulated through `LogitsProcessor` (and `LogitsWarper`) classes. Those classes are the code implementation behind flags like `temperature` or `top_k`. 

Most of our `LogitsProcessor` classes have a docstring that briefly describes their effect. However, unless you are an expert in text generation, it's hard to fully grasp the impact of using each class. In some cases, it is also non-trivial to prepare the arguments to initialize the `LogitsProcessor` class. As such, each class should have a clear usage example with `.generate()` in their docstring! 💪 

Here is an example: [SequenceBiasLogitsProcessor docstring](https://github.com/huggingface/transformers/blob/f1732e1374a082bf8e43bd0e4aa8a2da21a32a21/src/transformers/generation/logits_process.py#L559). Contrarily to the other classes (at the time of writing), we can quickly learn how to use it just by reading its docstring. We are also immediately aware of a few caveats 🤓 

Bonus points: our docstring examples are part of our CI, so we would be beefing up our tests to ensure we don't add regressions 🤗 

This issue is part of the [text generation docs rework](https://github.com/huggingface/transformers/issues/24575).

# How to participate?
1. Ensure you've read our contributing [guidelines](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md) 📜 
2. Claim your `LogitProcessor` class in this thread (confirm no one is working on it). You can check the full list of classes below, and you can find their implementation in [this file](https://github.com/huggingface/transformers/blob/main/src/transformers/generation/logits_process.py) 🎯 
    - You may need to do some detective work to fully understand the purpose of the class. For instance, some classes were created as part of a paper to be applied to any model, others are model-specific, and some exist to avoid weird bugs 🕵️ 
    - Looking at the git history is a great way to understand how a `LogitsProcessor` came to be.
3. Implement your changes, taking the [SequenceBiasLogitsProcessor docstring](https://github.com/huggingface/transformers/blob/f1732e1374a082bf8e43bd0e4aa8a2da21a32a21/src/transformers/generation/logits_process.py#L559) as reference  💪 
    - Add a clear example that calls the processor through `.generate()`. Make sure the example's outputs are correct and that the model used in the test is a small model (anything larger than GPT2 needs explicit approval);
    - If you feel like the original docstring could be better, feel free to enhance it as well!
    - Don't forget to run `make fixup` before your final commit.
4. Open the PR and tag me in it  🎊 

# Tracker
- [x] MinNewTokensLengthLogitsProcessor
- [x] TemperatureLogitsWarper
- [x] RepetitionPenaltyLogitsProcessor
- [ ] EncoderRepetitionPenaltyLogitsProcessor
- [x] TopPLogitsWarper
- [ ] TopKLogitsWarper
- [ ] TypicalLogitsWarper
- [x] EpsilonLogitsWarper
- [x] EtaLogitsWarper
- [x] NoRepeatNGramLogitsProcessor
- [ ] EncoderNoRepeatNGramLogitsProcessor
- [x] SequenceBiasLogitsProcessor
- [x] NoBadWordsLogitsProcessor
- [ ] PrefixConstrainedLogitsProcessor
- [x] HammingDiversityLogitsProcessor
- [ ] ForcedBOSTokenLogitsProcessor
- [ ] ForcedEOSTokenLogitsProcessor
- [ ] InfNanRemoveLogitsProcessor
- [ ] ExponentialDecayLengthPenalty
- [ ] LogitNormalization
- [ ] SuppressTokensAtBeginLogitsProcessor
- [ ] SuppressTokensLogitsProcessor
- [ ] ForceTokensLogitsProcessor
- [ ] WhisperTimeStampLogitsProcessor
- [ ] ClassifierFreeGuidanceLogitsProcessor

### PR DESCRIPTION
**Pull request by akshayamadhuri**
# What does this PR do?
Added some doc string to TypicalLogitsWarper with some examples as well.

Related to #24783 

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by akshayamadhuri** [comment]
@gante let me know the changes

File: src/transformers/generation/logits_process.py [hunk 1]
```diff
@@ -392,17 +392,57 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
 
 class TypicalLogitsWarper(LogitsWarper):
     r"""
-    [`LogitsWarper`] that performs typical decoding. See [Typical Decoding for Natural Language
-    Generation](https://arxiv.org/abs/2202.00666) for more information.
+    [`LogitsWarper`] that enforces the criterion of locally typical sampling when generating from probabilistic language models.
+     The abstraction of natural language generation as a discrete stochastic process, analyzed through information theory,
+    provides insights into the behavior of probabilistic language generators. This warping technique aims to produce more
```
Comment (gante) [inline src/transformers/generation/logits_process.py:397]:
```suggestion
     This warping technique aims to produce more
```

File: src/transformers/generation/logits_process.py [hunk 2]
```diff
@@ -392,17 +392,57 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
...
+    i.e., the conditional entropy of the model. See [Typical Decoding for Natural Language Generation](https://arxiv.org/abs/2202.00666) for more information.
 
+    <Tip>
+
+     The `TypicalLogitsWarper` is a special case of a `LogitsWarper`, which is itself a subclass of `LogitsProcessor`.
+        It is designed to modify the logits during text generation with probabilistic language models when `do_sample=True`.
+        By using `TypicalLogitsWarper`, you can enforce locally typical sampling, resulting in more coherent and fluent text,
+        while reducing degenerate repetitions.
+
+    </Tip>
```
Comment (gante) [inline src/transformers/generation/logits_process.py:408]: This `tip` could be summarized as a single short sentence, where it would point to the need to pass `do_sample=True` (without repeating what typical sampling is for, as it is already stated in at the top)

File: src/transformers/generation/logits_process.py [hunk 3]
```diff
@@ -392,17 +392,57 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
...
         filter_value (`float`, *optional*, defaults to `-float("Inf")`):
-            All filtered values will be set to this float value.
+            The value used to filter out logits that fall below this threshold. Any logits less than this value will be
+                set to -infinity before applying the softmax function. This helps in excluding unlikely tokens during sampling.
+                Default is -infinity.
         min_tokens_to_keep (`int`, *optional*, defaults to 1):
-            Minimum number of tokens that cannot be filtered.
-    """
+            The minimum number of tokens to always keep during sampling. This helps ensure a minimum level of coherence
+                in the generated text. Default is 1.
```
Comment (gante) [inline src/transformers/generation/logits_process.py:421]:
```suggestion
            The minimum number of tokens to always keep during sampling. The default is 1.
```

File: src/transformers/generation/logits_process.py [hunk 4]
```diff
@@ -392,17 +392,57 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
...
+         tokenizer = GPT2Tokenizer.from_pretrained(model_name)
+         model = GPT2LMHeadModel.from_pretrained(model_name)
+
+         # Create the warper with desired parameters
+         warper = TypicalLogitsWarper(mass=0.9, filter_value=-5.0, min_tokens_to_keep=2)
+
+         # Generate text using the model and warper
+         input_text = "Once upon a time"
+         input_ids = tokenizer.encode(input_text, return_tensors="pt")
+         output = model.generate(input_ids, do_sample=True, max_length=50, warper=warper)
```
Comment (gante) [inline src/transformers/generation/logits_process.py:438]: While technically correct, I'd prefer the example to set `TypicalLogitsWarper` indirectly through the `typical_p` argument (which is what most users would end up doing), as opposed to passing the warper directly [[docs](https://huggingface.co/docs/transformers/main/en/main_classes/text_generation#transformers.GenerationConfig.typical_p)]

File: src/transformers/generation/logits_process.py [hunk 5]
```diff
@@ -392,17 +392,57 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
...
+
+        # Decode and print the generated text
+        generated_text = tokenizer.decode(output[0], skip_special_tokens=True)
+        print("Generated Text: ", generated_text)
+     Once upon a time, there was a magical kingdom filled with mystical creatures and enchanted forests...
```
Comment (gante) [inline src/transformers/generation/logits_process.py:443]: Because this example relies on sampling, we will get a different output each time we run this script -- meaning that this example will fail our internal tests, which confirm that the examples are still valid :D 

A work-around is to set a seed [[docs](https://huggingface.co/docs/transformers/main/en/internal/trainer_utils#transformers.set_seed)]:
```py
from transformers import set_seed
set_seed(0)
```

**Review by gante — COMMENTED** [review]
Thank you for opening the PR 🙌 

A few nits to improve information density in the docs, and to improve the example

**Review by gante — COMMENTED** [review]
See the note in the example -- and don't forget to run `make fixup` on your end before the commit, otherwise our tests will block your PR! 🤗

File: src/transformers/generation/logits_process.py [hunk 6]
```diff
@@ -392,16 +392,58 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
...
+    ```python    
+    >>> from transformers import GPT2Tokenizer, GPT2LMHeadModel, TypicalLogitsWarper, set_seed
+
+    >>> # Load pre-trained model and tokenizer
+    >>> model_name = "gpt2"
+    >>> tokenizer = GPT2Tokenizer.from_pretrained(model_name)
+    >>> model = GPT2LMHeadModel.from_pretrained(model_name)
+
+    >>> # Set up the warper with desired parameters
+    >>> warper = TypicalLogitsWarper(tikohn_n=3, pi=0.95)
```
Comment (gante) [inline src/transformers/generation/logits_process.py:427]:
```suggestion
```

**Review by gante — APPROVED** [review]
Thank you for iterating :)

You still need to run `make fixup` and commit the resulting changes, CI is complaining about formatting

File: src/transformers/generation/logits_process.py [hunk 7]
```diff
@@ -392,16 +392,55 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
...
+      When using probabilistic language models for text generation, pass do_sample=True to enable locally typical sampling with reduced repetitions and improved coherence.
+    
+    </Tip>
+    
     Args:
         mass (`float`):
-            Value of typical_p between 0 and 1 inclusive, defaults to 0.9.
+             The proportion of probability mass to retain while warping the logits. The value should be between 0 and 1.
+                Higher values (close to 1.0) retain more probability mass, leading to more typical sampling, whereas lower
+                values (close to 0.0) retain less probability mass, leading to more diverse sampling. The default is 0.9.
```
Comment (amyeroberts) [inline src/transformers/generation/logits_process.py:408]: nit: We don't use the extra indent elsewhere for continued paragraphs

```suggestion
             The proportion of probability mass to retain while warping the logits. The value should be between 0 and 1.
             Higher values (close to 1.0) retain more probability mass, leading to more typical sampling, whereas lower
             values (close to 0.0) retain less probability mass, leading to more diverse sampling.
```

File: src/transformers/generation/logits_process.py [hunk 8]
```diff
@@ -392,16 +392,55 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
...
         filter_value (`float`, *optional*, defaults to `-float("Inf")`):
-            All filtered values will be set to this float value.
+             The value used to filter out logits that fall below this threshold. Any logits less than this value will be
+                set to -infinity before applying the softmax function. This helps in excluding unlikely tokens during sampling.
+                Default is -infinity.
```
Comment (amyeroberts) [inline src/transformers/generation/logits_process.py:412]: nit: same as above re indent. We don't need to say the default in the type info as `, defaults to ...` above 

```suggestion
             The value used to filter out logits that fall below this threshold. Any logits less than this value will be
             set to -infinity before applying the softmax function. This helps in excluding unlikely tokens during sampling.
```

File: src/transformers/generation/logits_process.py [hunk 9]
```diff
@@ -392,16 +392,55 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
...
         min_tokens_to_keep (`int`, *optional*, defaults to 1):
-            Minimum number of tokens that cannot be filtered.
+           The minimum number of tokens to always keep during sampling. The default is 1.
```
Comment (amyeroberts) [inline src/transformers/generation/logits_process.py:414]:
```suggestion
           The minimum number of tokens to always keep during sampling.
```

File: src/transformers/generation/logits_process.py [hunk 10]
```diff
@@ -392,16 +392,55 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
...
+
+    >>> # prompt
+    >>> input_prompt = "In today's meeting, the CEO discussed the company's future plans to"
+
+    >>> # Tokenize the sequence
+    >>> input_ids = tokenizer.encode(input_prompt, return_tensors="pt")
+
+    >>> # Generate text using the model and warper
+    >>> typical_p = 0.9  
+    >>> output = model.generate(input_ids, do_sample=True, max_length=50, typical_p=typical_p)
```
Comment (amyeroberts) [inline src/transformers/generation/logits_process.py:437]: It would be nice (but not necessary) to have two examples here with high and low typical_p values to demonstrate its effects

**Review by amyeroberts — APPROVED** [review]
Thanks for adding this! 🤗 

Just some nits on the formatting
